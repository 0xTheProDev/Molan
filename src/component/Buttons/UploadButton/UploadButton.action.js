// @flow
import {
  InputElementRefObject,
  FileInputChangeHandler,
  FileInputErrorHandler,
} from './UploadButton.flow';

export function onFileButtonClick(refObject: InputElementRefObject): void {
  return function onClickHandler(event: MouseEvent) {
    if (refObject.current) {
      refObject.current.click();
    }
  }
}

export function onFileInputChange(onUploadFile: FileInputChangeHandler, onUploadFileError: FileInputErrorHandler) {
  return function onChangeHandler(event: InputEvent): void {
    const files: FileList = event.target.files;

    if (!files || files.length !== 1) {
      onUploadFileError();
      return;
    }

    const fileReader = new FileReader();

    fileReader.addEventListener('load', (ev: ProgressEvent<FileReader>) => {
      onUploadFile(ev.target.result);
    });
    fileReader.addEventListener('error', (ev: ProgressEvent<FileReader>) => {
      onUploadFileError(ev.target.error);
    });

    fileReader.readAsText(files[0], 'utf-8');
  };
}
