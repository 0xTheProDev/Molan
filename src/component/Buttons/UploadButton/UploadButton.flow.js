// @flow
export type UploadButtonProps = {
  UploadButtonText: string,
  onUploadFile: FileInputChangeHandler,
  onUploadFileError: FileInputErrorHandler,
};

export type UploadButtonInternalProps = {
  AllowedFileTypes: string,
  InputElementRef: InputElementRefObject,
  UploadButtonText: string,
  onChangeHandler: (event: InputEvent) => void,
  onClickHandler: (event: MouseEvent) => void,
};

export type InputElementRefObject = {
  current?: HTMLInputElement,
};

export type FileInputChangeHandler = (fileContent: string) => void;

export type FileInputErrorHandler = (error?: string) => void;
