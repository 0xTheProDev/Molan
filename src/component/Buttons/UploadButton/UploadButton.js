// @flow
import React from 'react';
import { Button, Icon } from 'semantic-ui-react';
import { noop } from 'util/common';

import { UploadButtonInternalProps } from './UploadButton.flow';
import Styles from './UploadButton.module.css';

export function UploadButton(props: UploadButtonInternalProps) {
  const {
    InputElementRef,
    UploadButtonText,
    AllowedFileTypes,
    onClickHandler,
    onChangeHandler,
  } = props;

  return (
    <Button
      animated="vertical"
      floated="right"
      title={UploadButtonText}
      onClick={onClickHandler}
    >
      <Button.Content hidden>{UploadButtonText}</Button.Content>
      <Button.Content visible>
        <Icon name="upload" />
      </Button.Content>
      <input
        ref={InputElementRef}
        type="file"
        accepts={AllowedFileTypes}
        onChange={onChangeHandler}
        className={Styles['file-input']}
      />
    </Button>
  );
}

UploadButton.defaultProps = {
  UploadButtonText: 'Upload',
  onChangeHandler: noop,
  onClickHandler: noop,
};
