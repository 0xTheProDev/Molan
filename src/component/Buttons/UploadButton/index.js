// @flow
import React from 'react';
import { MIME_TYPES } from 'constants/templates';

import { UploadButton } from './UploadButton';
import { UploadButtonProps } from './UploadButton.flow';
import { onFileButtonClick, onFileInputChange } from './UploadButton.action';

export default function UploadButtonFactory(props: UploadButtonProps) {
  const {
    UploadButtonText,
    onUploadFile,
    onUploadFileError,
  } = props;

  const AllowedMimeTypes = new Set(Object.values(MIME_TYPES));
  const AllowedFileTypes = Array.from(AllowedMimeTypes).join(',');

  const InputElementRef = React.createRef();

  const onClickHandler = onFileButtonClick(InputElementRef);
  const onChangeHandler = onFileInputChange(onUploadFile, onUploadFileError);

  return (
    <UploadButton
      AllowedFileTypes={AllowedFileTypes}
      InputElementRef={InputElementRef}
      UploadButtonText={UploadButtonText}
      onClickHandler={onClickHandler}
      onChangeHandler={onChangeHandler}
    />
  );
}

export { UploadButtonProps, UploadButtonFactory as UploadButton };
