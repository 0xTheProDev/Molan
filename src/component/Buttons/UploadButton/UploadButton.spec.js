import React from 'react';
import { Button } from 'semantic-ui-react';

import { UploadButton } from './UploadButton';

describe('Component::Buttons:UploadButton::UploadButton', () => {
  let
    onChangeHandlerStub,
    onClickHandlerStub,
    refObjectStub,
    minProps;

  beforeAll(() => {
    onChangeHandlerStub = jest.fn();
    onClickHandlerStub = jest.fn();

    refObjectStub = React.createRef();

    minProps = {
      AllowedFileTypes: 'text/plain',
      InputElementRef: refObjectStub,
      UploadButtonText: 'Upload',
      onChangeHandler: onChangeHandlerStub,
      onClickHandler: onClickHandlerStub,
    }; 
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render with base structure', () => {
    const wrapper = shallow(<UploadButton {...minProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should store reference of InputElement', () => {
    mount(<UploadButton {...minProps} />);
    expect(refObjectStub.current).toBeTruthy();
  });

  it('should call onClickHandler when button is clicked', () => {
    const wrapper = shallow(<UploadButton {...minProps} />);

    wrapper.find(Button).simulate('click');
    expect(onClickHandlerStub).toHaveBeenCalled();
  });

  it('should call onChange handler when file input value changes', () => {
    const wrapper = shallow(<UploadButton {...minProps} />);

    wrapper.find('input').simulate('change');
    expect(onChangeHandlerStub).toHaveBeenCalled();
  });
});
