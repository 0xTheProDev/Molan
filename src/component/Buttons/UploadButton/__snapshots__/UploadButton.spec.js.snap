// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component::Buttons:UploadButton::UploadButton should render with base structure 1`] = `
<Button
  animated="vertical"
  as="button"
  floated="right"
  onClick={[MockFunction]}
  title="Upload"
>
  <ButtonContent
    hidden={true}
  >
    Upload
  </ButtonContent>
  <ButtonContent
    visible={true}
  >
    <Icon
      as="i"
      name="upload"
    />
  </ButtonContent>
  <input
    accepts="text/plain"
    className="file-input"
    onChange={[MockFunction]}
    type="file"
  />
</Button>
`;
